source/dunders/index.rst

```diff
@@ -1,6 +1,184 @@
+          }
+
+          def __init__(self, name) :
+              self.name=name
+
+      sc = SomeClass("Totoro")
+      print (len(sc.name)) #affiche la taille du string (6)
+      print (len(sc.grammar)) #affiche le nombre d'éléments
+
+
+Une autre grande utilisation des dunders consiste à les surcharger de manière à les personalisé.
+
+De plus, par convention, on déclare une méthode privée comme étant un dunder :
+
+.. code-block:: python3
+
+      def __myPrivateMethod__(self, other):
+          return none;
+
+
+Exemples
+========
+
+.. code-block:: python3
+
+      from math import hypot
+
+      class Vector:
+          def __init__(self, x=0, y=0):
+              self.x = x
+              self.y = y
+
+          def __repr__(self):
+              return 'Vecteur(%r, %r)' % (self.x, self.y)
+
+          def __abs__(self):
+               return hypot(self.x, self.y)
+
+          def __add__(self, other):
+              #Overide de l'opérateur +
+              x = self.x + other.x
+              y = self.y + other.y
+              return Vector(x, y)
+
+          def __mul__(self, scalar):
+              return Vector(self.x * scalar, self.y * scalar)
+
+          def __truediv__(self, other):
+              raise TypeError("On ne peut pas diviser un vecteur !")
+
+      vect = Vector(4,7)
+      print(repr(vect)+" a pour norme "+str(abs(vect))) #Vecteur(4,7) a pour norme 8.0622
+      vect = vect*2 #Vecteur(8,14)
+      vect+=Vector(2,5)  #Vecteur(10,19)
+      vect/=5 #lève une exception
+
+
+.. code-block:: python3
+
+      class Main(object):
+
+      def __init__(self, *args):
+        self.cartes = args
+
+      def ajouter(self, carte):
+        obj2 = list(self.cartes) #Converti en liste
+        obj2.append(carte)
+        self.cartes = tuple(obj2)
+
+      def __str__(self):
+        return str(u'; '.join(self.cartes).encode('utf8'))
+
+      def __len__(self):
+        return len(self.cartes)
+
+
+      def __getitem__(self, key):
+        #Est appelé quand on fait objet[index] ou objet[key].
+        return self.cartes[key]
+
+      def __iter__(self):
+        #Est appelé quand on fait un iter(objet). La valeur retournée doit être un iterateur.
+        return iter(self.cartes)
+
+
+      def __reversed__(self):
+        #Est appelé quand on fait reversed(objt)
+        return reversed(self.cartes)
+
+
+      def __contains__(self, item):
+        #Est appelé quand "in objet"
+        return item in self.cartes
+
+      main = Main('1Coeur', '7Pique')
+      print(str(main))
+
+      main.ajouter('AsCoeur')
+
+      for carte in main: # parce qu'on a défini __iter__ !
+      print (carte)
+
+      print (main[0]) # 1Coeur
+
+      print ('3Coeur' in main) # False
+
+      print (len(main))# 3
+
+
+Différence entre str et repr :
+
+.. code-block:: python3
+
+        print(str(3)==str("3")) # return True car str est ambigu
+        print(repr(3)==repr("3")) # return False car non-ambigu (Int != String)
+
```

Comment:
Vous pouvez faire mieux. Par exemple

http://stackoverflow.com/questions/1436703/difference-between-str-and-repr-in-python#2626364